src/windows/list.cpp

```diff
@@ -0,0 +1,687 @@
+    &dataType, (LPBYTE) buffer, sizeof(buffer), &size);
+
+  if (!hasHardwareId) {
+    return false;
+  }
+
+  // printf("SPDRP_HARDWAREID: %s\n", buffer);
+
+  std::string hardwareId(buffer);
+
+  for (std::string vhdHardwareId : VHD_HARDWARE_IDS) {
+    if (hardwareId.find(vhdHardwareId, 0) != std::string::npos) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool IsSystemDevice(HDEVINFO hDeviceInfo, SP_DEVINFO_DATA deviceInfoData) {
+  PWSTR windowsPath = NULL;
+  HRESULT result = SHGetKnownFolderPath(
+    FOLDERID_Windows, 0, NULL, &windowsPath);
+
+  if (result == S_OK) {
+    std::string systemPath = WCharToUtf8(windowsPath);
+    CoTaskMemFree(windowsPath);
+    printf("systemPath %s\n", systemPath.c_str());
+    // TODO(jhermsmeier): Compare against mountpoints to actually determine this
+    return true;
+  }
+
+  CoTaskMemFree(windowsPath);
+  return false;
+}
+
+std::vector<std::string> GetAvailableVolumes() {
+  DWORD logicalDrivesMask = GetLogicalDrives();
+  std::vector<std::string> logicalDrives;
+
+  if (logicalDrivesMask == 0) {
+    return logicalDrives;
+  }
+
+  char currentDriveLetter = 'A';
+
+  while (logicalDrivesMask) {
+    if (logicalDrivesMask & 1) {
+      logicalDrives.push_back(std::string(1, currentDriveLetter));
+    }
+    currentDriveLetter++;
+    logicalDrivesMask >>= 1;
+  }
+
+  return logicalDrives;
+}
+
+int32_t GetDeviceNumber(HANDLE hDevice) {
+  BOOL result;
+  DWORD size;
+  DWORD errorCode = 0;
+  int32_t diskNumber = -1;
+
+  STORAGE_DEVICE_NUMBER deviceNumber;
+  VOLUME_DISK_EXTENTS diskExtents;
+
+  // Some devices will have the diskNumber exposed through their disk extents,
+  // while most of them will only have one accessible through
+  // `IOCTL_STORAGE_GET_DEVICE_NUMBER`, so we check this one first,
+  // augmenting / overriding it with the latter
+  result = DeviceIoControl(
+    hDevice, IOCTL_VOLUME_GET_VOLUME_DISK_EXTENTS, NULL, 0,
+    &diskExtents, sizeof(VOLUME_DISK_EXTENTS), &size, NULL);
+
+  if (result && diskExtents.NumberOfDiskExtents > 0) {
+    printf("[INFO] DiskNumber: %i\n", diskExtents.Extents[0].DiskNumber);
+    // NOTE: Always ignore RAIDs
+    if (diskExtents.NumberOfDiskExtents >= 2) {
+      printf("[INFO] Possible RAID: %i\n", diskExtents.Extents[0].DiskNumber);
+      return -1;
+    }
+    diskNumber = diskExtents.Extents[0].DiskNumber;
+  } else {
+    errorCode = GetLastError();
+    printf("[INFO] VOLUME_GET_VOLUME_DISK_EXTENTS: Error 0x%08lX\n", errorCode);
+  }
+
+  result = DeviceIoControl(
+    hDevice, IOCTL_STORAGE_GET_DEVICE_NUMBER, NULL, 0,
+    &deviceNumber, sizeof(deviceNumber), &size, NULL);
+
+  if (result) {
+    printf("[INFO] DeviceNumber: %i\n", deviceNumber.DeviceNumber);
+    diskNumber = deviceNumber.DeviceNumber;
+  } else {
+    errorCode = GetLastError();
+    printf("[INFO] STORAGE_GET_DEVICE_NUMBER: Error 0x%08lX\n", errorCode);
+  }
+
+  return diskNumber;
+}
+
+void GetMountpoints(int32_t deviceNumber,
```

Comment:
Folder-mapped volumes aren't drives or storage devices; they're just mapped paths and since we're enumerating through storage drivers here, they don't even show up (which is what we want).

Reply:
I guess the only possibly-minor-inconsistency here is that if (a partition on) drive1 is mounted to a drive-letter, and a partition on drive2 is mounted to just a folder-path, then the `mountpoints` property would show drive1 as having a mountpoint, but drive2 as having no mountpoints. (so a user might assume that drive2 isn't "in use", when it actually is)

Although this isn't something I've experimented with much myself, so perhaps I'm misunderstanding some of this. I did a quick google and found e.g. https://technet.microsoft.com/en-us/library/cc753321(v=ws.11).aspx which says "assign a mount-point folder path (rather than a drive letter) to the drive".

Reply:
I think you're confusing that with the subst-mapped drives many users had a problem with –
 which goes the other direction, mapping a folder -> drive letter, rather than drive -> NTFS subfolder. The latter case doesn't really matter, as the storage device will still be correctly enumerated.

Reply:
> I think you're confusing that with the subst-mapped drives

Nope ;-)

> The latter case doesn't really matter, as the storage device will still be correctly enumerated.

...with the only minor 'wrinkle' being that (AIUI) a drive mapped to a NTFS folder won't show any entries in the `mountpoints` property, as I mentioned above. I don't disagree that it'll still be correctly enumerated.